### PR TITLE
AccessLoggingPolicy Design

### PR DESCRIPTION
**Pull request by zirain**
xref: https://github.com/envoyproxy/gateway/issues/280

### CODE REVIEW

**Comment by arkodg** [comment]
@zirain thanks for designing this API !
Hoping this feature can be added to v0.5.0 as part of https://github.com/envoyproxy/gateway/issues/699

For the time being (in v0.4.0) the user could directly configure the fields using a custom bootstrap

**Comment by zirain** [comment]
> @zirain thanks for designing this API ! Hoping this feature can be added to v0.5.0 as part of #699
> 
> For the time being (in v0.4.0) the user could directly configure the fields using a custom bootstrap

yes, my goal is releasing this in v5.0.0-rc1, plz take a review when you're free.

**Review by kflynn — APPROVED** [review]
Looks good! 🙂

**Review by AliceProxy — APPROVED** [review]
Design looks good to me. The comments and examples are very clear. It would be awesome to split the design doc into a user-doc for examples and add some of that context from the comments if you wanted to. :slightly_smiling_face:

**Comment by zirain** [comment]
@arkodg can you take a look?

one more key point: should add `listenerMatch`, which will allow user set different provider per listener.

**Comment by arkodg** [comment]
> @arkodg can you take a look?
> 
> one more key point: should add `listenerMatch`, which will allow user set different provider per listener.

I prefer having someone own the overall observability epic first, understand what all we plan on support, and for which user type and then including this access logging API

**Comment by zirain** [comment]
AcessLogging: this PR
Tracing: I had a rough idea in my mind, will write a design doc ASAP.
Metrics: there's a lot of bulit-in stats in envoy, does EG need do custom things like istio done?

**Comment by arkodg** [comment]
> AcessLogging: this PR
> Tracing: I had a rough idea in my mind, will write a design doc ASAP.
> Metrics: there's a lot of bulit-in stats in envoy, does EG need do custom things like istio done?

We would also need higher level design, e.g.  what kind of observability features does an `app dev` need to run apps in production successfully as well as operate EG, and what does EG provide and integrate with to make it happen

**Comment by haq204** [comment]
> AcessLogging: this PR Tracing: I had a rough idea in my mind, will write a design doc ASAP. Metrics: there's a lot of bulit-in stats in envoy, does EG need do custom things like istio done?

For metrics I think it would be helpful to have some control plane metrics (e.g translation time for each runner)

I have a commit in my fork that added some prom metrics that I can pitch on that front

**Comment by zirain** [comment]
>  what kind of observability features does an app dev need to run apps in production successfully as well as operate EG, and what does EG provide and integrate with to make it happen

can guys from contour/Ambassador share with us?

**Comment by arkodg** [comment]
we are currently adding aux control plane config in the [EnvoyGateway](https://github.com/envoyproxy/gateway/blob/main/api/config/v1alpha1/envoygateway_types.go) resource, and any data plane config (like logging) in the[ EnvoyProxy](https://github.com/envoyproxy/gateway/blob/384577334dea2c0ba28800643a2b94fea4eed244/api/config/v1alpha1/envoyproxy_types.go#L40) resource, prefer if we didnt have to create another CR .

**Comment by zirain** [comment]
AccessLog is part of xds, do you means add things to `EnvoyProxy` CR?

**Comment by arkodg** [comment]
> AccessLog is part of xds, do you means add things to `EnvoyProxy` CR?

yeah although https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#access-logging is part of HCM, imo its something that the platform admin persona will setup once (using the `EnvoyProxy` CR)

**Comment by zirain** [comment]
> > AccessLog is part of xds, do you means add things to `EnvoyProxy` CR?
> 
> yeah although https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#access-logging is part of HCM, imo its something that the platform admin persona will setup once (using the `EnvoyProxy` CR)

If I understand right, change `EnvoyProxy` will cause a restart of dataplane? if so, prefer not put things there.

**Comment by arkodg** [comment]
> > > AccessLog is part of xds, do you means add things to `EnvoyProxy` CR?
> > 
> > 
> > yeah although https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#access-logging is part of HCM, imo its something that the platform admin persona will setup once (using the `EnvoyProxy` CR)
> 
> If I understand right, change `EnvoyProxy` will cause a restart of dataplane? if so, prefer not put things there.

for this case it shouldnt, since it will result in a xds push

**Comment by arkodg** [comment]
thoughts on having 
```
telemetry:
   log:
   metrics:
   traces:
   collector:
```   
In EnvoyProxy CR (for data plane obs) and one in EnvoyGateway CR (for control plane obs)

**Comment by zirain** [comment]
For now, if things in [EnvoyProxy](https://github.com/envoyproxy/gateway/blob/main/api/config/v1alpha1/envoyproxy_types.go) changes, it will trigger a rollout of  data plane deployment. it's simple and easy to understand. If we put observability there, EG need find out which fields are changed. 

> moving the config into the EnvoyGateway CRD

if we agree that EG won't support custom observability(accesslog,tracing) per gateway/listener, it's ok to me.

**Comment by arkodg** [comment]
> For now, if things in [EnvoyProxy](https://github.com/envoyproxy/gateway/blob/main/api/config/v1alpha1/envoyproxy_types.go) changes, it will trigger a rollout of data plane deployment. it's simple and easy to understand. If we put observability there, EG need find out which fields are changed.
> 
> > moving the config into the EnvoyGateway CRD
> 
> if we agree that EG won't support custom observability(accesslog,tracing) per gateway/listener, it's ok to me.

* The infra manager is clever enough today to only update infra spec if some spec has changed, so unless some infra spec changes, it will not trigger a restart
* imo the common case is to set a consistent telemetry setting per gatewayClass by the admin. We could support custom cases with #24

**Comment by zirain** [comment]
From my experience, merge does not work as expected on slice type 
https://github.com/envoyproxy/go-control-plane/blob/bf4f5e3fc9695e5b65b671a0ebd217ae09ba285e/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.pb.go#L490

**Comment by zirain** [comment]
hold this before https://github.com/envoyproxy/gateway/issues/701

File: docs/latest/design/accesslogging.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+Envoy supports extensible access logging to different sinks, File, gRPC etc.
+Envoy supports customizable access log formats using predefined fields as well as arbitrary HTTP request and response headers.
+Envoy supports several built-in access log filters and extension filters that are registered at runtime.
+EG should provide an easy to config access log formats and sinks per `gateway`, 
+base on the `AcessLoggingPolicy` API type proposed in this design document.
+
+Envoy Gateway leverages [Gateway API][] for configuring managed Envoy proxies. Gateway API defines core, extended, and
+implementation-specific API [support levels][] for implementors such as Envoy Gateway to expose features. Since
+access logging is not covered by `Core` or `Extended` APIs, an [Policy Attachment][] API will
```
Comment (Xunzhuo) [inline docs/latest/design/accesslogging.md:13]: Policy Attachment ot extensionRef ?

Comment (Xunzhuo) [reply docs/latest/design/accesslogging.md:13]: I see, it makes sense.

**Review by qicz — APPROVED** [review]
Great Job

**Comment by zirain** [comment]
replaced by https://github.com/envoyproxy/gateway/pull/1407

**Pull request closed without merging**